src/Text/Microstache/Type.hs

```diff
@@ -126,3 +128,20 @@ instance Exception MustacheException where
 #else
 instance Exception MustacheException
 #endif
+
+data MustacheWarning
+  = MustacheVariableNotFound Key
+    -- ^ The template contained a variable for which there was no data counterpart in the current context
+  | MustacheDirectlyRenderedValue Key
+    -- ^ A complex value such as an Object or Array was directly rendered into the template
```

Comment:
@RyanGlScott , I'm happy to change an API so adding new warnings would be non-breaking change,
I wonder is someone would need exhaustive matching on warnings (as that will break any time one adds a warning).

With product types it's easy: smart constructors and field accessors/lens. is there good ways for sums?

---

- I found a place where to check for `PartialNotFound`, I'l try to write a test possibly triggering it.
- `SectionTargetNotFound` is subsumed by `MustacheVariableNotFound`
    - I can split it, if that's a good idea?
- `InvalidImplicitSectionContextType` is when

```haskell
>>> let Right templ = compileMustacheText "" "{{#string}} {{.}} {{#.}} ?  {{/.}} {{/string}}"
>>> renderMustache templ $ object [ "string" .= ("foobar" :: Text) ]
" foobar  ?   "
```

```
% cat ex1.json 
{ "string": "foobar" }
% cat ex.mustache 
{{#string}} {{.}} {{#.}} ?  {{/.}} {{/string}}
% ./node_modules/.bin/mustache ex1.json ex.mustache
 foobar  ?   
```

when given `{ string: "string" }`.

Quite obscure, I'm not really sure that something I'd bother to check.

For something like auto-flattening?

```
>>> let Right templ = compileMustacheText "" "{{#foos}}{{#.}}{{#.}}{{.}};{{/.}}{{/.}}{{/foos}}"
>>> renderMustache templ $ object [ "foos" .= ["foo1" :: Text, "foo2"] ]
"foo1;foo2;"
>>> renderMustache templ $ object [ "foos" .= [["foo1" :: Text, "foo2"], ["foo3"]] ]
"foo1;foo2;foo3;"
```

In the first case `mustache` would error, as inner `{{#.}}` is pointing to json string.